### ISSUE TITLE
username/password silently dropped from durable object request urls

### ISSUE THREAD

**Issue by jed** [cloudflare/workerd#194]
Given the following worker:
```js
export default {
  fetch(req, env) {
    let url = new URL(req.url)
    url.username = 'kentonv'
    return env.do.get(env.do.newUniqueId()).fetch(url, req)
  }
}

export class DO {
  fetch(req) {
    return new Response(req.url)
  }
}
```
The URL returned by the durable object's response does not contain the `username` set in the worker, despite serializing correctly (ie, shows up in `url.href`) inside the worker itself.

It seems as though `username` and `password` are being silently dropped, despite being valid URL components. Ideally URLs would be sent as-is, but if for some unfortunate reason workerd strips them intentionally, it fetch should fail in the worker itself before reaching the DO (same as protocol errors a la `fetch_refuses_unknown_protocols`).

**Comment by kentonv** [comment]
Hmm, I'm not sure. If you perform `fetch()` in a browser and gave it a URL with username/password, the username and password do not remain in the URL as seen on the server. Instead, the `Authorization` header is filled per HTTP basic auth. In fact, it wouldn't be possible to keep them in the URL since the HTTP protocol doesn't actually transmit the full URL as a single string, it breaks it up and puts all the parts in various places.

I would argue that it is a bug that we don't fill in `Authorization`. (Assuming the `fetch()` spec does in fact say we should, at least...)

If we did fill that in, would that help with your use case?

**Comment by kentonv** [comment]
There's also an argument to be made that we should let you pass any string you want to `fetch()` when it's a worker-to-worker or worker-to-DO call, and that string should just show up unchanged on the receiving end. But I think we tried to implement that once and ran into too many problems.

**Comment by jed** [comment]
I think adding an `Authorization` header would be fine. Personally I'd avoid arbitrary strings, since this would cause the already disparate DO and worker `fetch` arguments to further diverge from each other... If you want to relax the HTTP format, which I think is a good idea, I'd add another method on the DO altogether that looks more like `postMessage`.

FWIW, here's what the spec says:
<img width="736" alt="image" src="https://user-images.githubusercontent.com/4433/204833381-b4d43e3c-2dc8-4ee6-80ad-44d2371fa226.png">

**Comment by jed** [comment]
This flag doesn't appear to be implemented in any browser I can find. `fetching` a URL like the above instead results in the following error, even when `credentials` is set to `include`: `TypeError: Failed to execute 'fetch' on 'Window': Request cannot be constructed from a URL that includes credentials`.

**Comment by jasnell** [comment]
Honestly I prefer the brower's behavior here. We really should not be encouraging putting the username/password into URLs and should instead encourage developers to set the `Authorization` header explicitly. 

Specifically, I'd prefer that this reject/throw: `fetch('https://jasnell@example.org')`

**Comment by kentonv** [comment]
FWIW I think `credentials = "include"` is meant to tell the browser to use the _ambient_ credentials, i.e. whatever the user has previously entered when prompted to log into the target site using HTTP basic auth. Of course, no one actually uses HTTP basic auth anymore and I think it's officially deprecated, even. But it makes sense to me that browser makers did not want to figure out how to reconcile ambient credentials with credentials specified in the URL.

Of course, in Workers, we don't have ambient credentials at all. But I agree we should do what the browsers do and throw an exception. We'll need a compat flag, of course.

**Comment by jed** [comment]
Agreed, throwing is probably the safest bet (with a flag a la `fetch_refuses_unknown_protocols`, as per my original suggestion).

(HTTP basic auth is definitely not deprecated though, just support for `username:password@` in URLs.)

**Comment by jasnell** [comment]
Ok, specifically, this is tied to the fact that our fetch implementation is really not compliant with the spec, so we'll include the change here within that bucket. Specifically, if we look at the definition of the `new Request(...)` constructor, [step 5](https://fetch.spec.whatwg.org/#dom-request) requires that we parse the input URL and throw if the parsed URL includes credentials.

![image](https://user-images.githubusercontent.com/439929/206285443-b7423109-1c4f-4507-8f5b-6c2bd0391938.png)

In our current implementation of `Request`, we do not parse the input URL at all until immediately before the fetch impl is called.

**Comment by kentonv** [comment]
@jasnell Sounds like you're saying there are two issues:

1. We parse at `fetch(request)` time and not `new Request()` time. (But this only makes a difference if you do `fetch(url)` without explicitly creating a `Request` object.)
2. When we do parse the URL (at `fetch()` time), we silently ignore credentials, when we should be throwing if they are present.

I could imagine these being two separate fixes with separate compat flags (if we care to fix them).

**Comment by kentonv** [comment]
> it would be nice if new URL(request.url) in a worker would also take into account the basic credentials from authorisation header, if provided

Hmm, I don't think that's possible without deviating from the spec -- `request.url` does not contain credentials. We'd have to introduce some sort of method like `request.getUrlWithCredentials()` but that would be non-standard.
